pkg/controlsvc/controlsvc_test.go

```diff
@@ -0,0 +1,647 @@
+
+	bridgeConnTestCases := []struct {
+		name          string
+		message       string
+		expectedCalls func()
+		expectedError bool
+		errorMessage  string
+	}{
+		{
+			name:    "without message and copier error",
+			message: "",
+			expectedCalls: func() {
+				mockCopier.EXPECT().Copy(gomock.Any(), gomock.Any()).Return(int64(0), errors.New("read from conn copy error"))
+			},
+			expectedError: true,
+			errorMessage:  "read from conn copy error",
+		},
+		{
+			name:    "with message and no error",
+			message: "message",
+			expectedCalls: func() {
+				mockCon.EXPECT().Write(gomock.Any()).Return(0, errors.New("read from conn write error"))
+			},
+			expectedError: true,
+			errorMessage:  "read from conn write error",
+		},
+		{
+			name:    "without message and no copier error",
+			message: "",
+			expectedCalls: func() {
+				mockCopier.EXPECT().Copy(gomock.Any(), gomock.Any()).Return(int64(0), nil)
+			},
+			expectedError: false,
+			errorMessage:  "",
+		},
+	}
+
+	for _, testCase := range bridgeConnTestCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			testCase.expectedCalls()
+			err := sockControl.ReadFromConn(testCase.message, mockCon)
+
+			if testCase.expectedError {
+				if err == nil && err.Error() != testCase.errorMessage {
+					printExpectedError(t, err)
+				}
+			} else {
+				if err != nil {
+					printExpectedError(t, err)
+				}
+			}
+		})
+	}
+}
+
+func TestSockControlWriteToConn(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mockCon := mock_controlsvc.NewMockConn(ctrl)
+	mockUtil := mock_controlsvc.NewMockUtiler(ctrl)
+	mockCopier := mock_controlsvc.NewMockCopier(ctrl)
+
+	sockControl := controlsvc.NewSockControl(mockCon, mockUtil, mockCopier)
+
+	bridgeConnTestCases := []struct {
+		name          string
+		message       string
+		input         chan []byte
+		expectedCalls func()
+		expectedError bool
+		errorMessage  string
+	}{
+		{
+			name:    "without message and with error",
+			message: "",
+			expectedCalls: func() {
+				mockCon.EXPECT().Write(gomock.Any()).Return(0, errors.New("write to conn chan error"))
+			},
+			expectedError: true,
+			errorMessage:  "write to conn chan error",
+		},
+		{
+			name:    "with message and with error",
+			message: "message",
+			expectedCalls: func() {
+				mockCon.EXPECT().Write(gomock.Any()).Return(0, errors.New(writeToConnError))
+			},
+			expectedError: true,
+			errorMessage:  writeToConnError,
+		},
+		{
+			name:    "without message and error",
+			message: "",
+			expectedCalls: func() {
+				mockCon.EXPECT().Write(gomock.Any()).Return(0, nil)
+			},
+			expectedError: false,
+			errorMessage:  writeToConnError,
+		},
+	}
+
+	for _, testCase := range bridgeConnTestCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			testCase.expectedCalls()
+			c := make(chan []byte)
+			go func(c chan []byte) {
+				c <- []byte{7}
+				defer close(c)
+			}(c)
+
+			err := sockControl.WriteToConn(testCase.message, c)
+
+			if testCase.expectedError {
```

Comment:
Same as the previous test, i dont understand this `if` block.

How can `err == nill` AND `err.Error()` exist?